python/python/bystro/supervised_ppca/_misc_pt.py

```diff
@@ -0,0 +1,287 @@
+
+    return Sigma_inv
+
+
+def ldet_sw_factor_analysis(Lambda, W, I_l=None):
+    """
+    Compute the log determinant of the covariance matrix in a factor 
+    analysis model using the Sherman Woodbury identity
+
+    Parameters
+    ----------
+    Lambda : torch.Tensor
+        Precision matrix in factor analysis.
+
+    W : torch.Tensor
+        Factor loading matrix.
+
+    I_l : torch.Tensor, optional
+        LxL identity matrix. If not provided, it is set to the identity matrix.
+
+    Returns
+    -------
+    torch.Tensor
+        Log determinant of the precision matrix.
+    """
+    if I_l is None:
+        I_l = torch.eye(W.shape[0])
+    ldet_L = torch.logdet(Lambda)
+    LiW = la.solve(Lambda, torch.transpose(W, 0, 1))
+    WtLiW = torch.matmul(W, LiW)
+    IWtLiW = I_l + WtLiW
+    ldet_p = torch.logdet(IWtLiW)
+    log_determinant = ldet_p + ldet_L
+    return log_determinant
+
+
+def inv_sw_factor_analysis(Lambda, W, I_l=None, I_p=None):
+    """
+    Compute the inverse of the covariance matrix in a factor analysis model.
+
+    Parameters
+    ----------
+    Lambda : torch.Tensor
+        Precision matrix in factor analysis.
+
+    W : torch.Tensor
+        Factor loading matrix.
+
+    I_l : torch.Tensor, optional
+        LxL identity matrix. If not provided, it is set to the identity matrix.
+
+    I_p : torch.Tensor, optional
+        PxP identity matrix. If not provided, it is set to the identity matrix.
+
+    Returns
+    -------
+    torch.Tensor
+        Inverse of the covariance matrix.
+    """
+    if I_l is None:
+        I_l = torch.eye(W.shape[0])
+    if I_p is None:
+        I_p = torch.eye(W.shape[1])
+
+    Lambda_inv = torch.inverse(Lambda)
+    WLi = torch.matmul(W, Lambda_inv)
+    inner = I_l + torch.matmul(WLi, torch.transpose(W, 0, 1))
+    inner_inv = la.inv(inner)
+    end = torch.matmul(inner_inv, WLi)
+    term2 = torch.matmul(torch.transpose(W, 0, 1), end)
+    Imterm2 = I_p - term2
+    Sigma_inv = torch.matmul(Lambda_inv, Imterm2)
+    return Sigma_inv
+
+
+def mvn_log_prob_sw(X, mu, Lambda, W, I_l=None):
+    """
+    Compute the log probability of data points under a multivariate 
+    normal distribution using the Sherman Woodbury matrix identity.
+
+    Parameters
+    ----------
+    X : torch.Tensor
+        Data points (rows are observations, columns are variables).
+
+    mu : torch.Tensor
+        Mean vector of the multivariate normal distribution.
+
+    Lambda : torch.Tensor
+        Precision matrix of the multivariate normal distribution.
+
+    W : torch.Tensor
+        Weight matrix.
+
+    I_l : torch.Tensor, optional
+        LxL identity matrix. If not provided, it is set to the 
+        identity matrix.
+
+    Returns
+    -------
+    torch.Tensor
+        Log probability of the data points under the multivariate 
+        normal distribution.
+    """
+    L, p = W.shape
+
+    if I_l is None:
+        I_l = torch.eye(L)
+
+    term1 = -p * 0.9189385332046727
```

Comment:
what is the purpose of multiplying by .918?

Reply:
Saves us the time of computing torch.log(2*torch.pi), since it doesn't even matter in optimization anyways.